consumer/batching.go

```diff
@@ -3,77 +3,276 @@ package consumer
 import (
 	"fmt"
 	"github.com/deso-protocol/core/lib"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"math"
+	"time"
 )
 
-type BatchedEntry struct {
-	Encoder *lib.DeSoEncoder
-	Key     []byte
+const (
+	retryLimit = 10
+)
+
+// BatchIndex is a struct that contains the index of the first entry in a batch, and the batch's index relative to all other executed batches.
+type BatchIndexInfo struct {
+	MinEntryIndex uint64
+	Index         uint64
+}
+
+// manageBatchedEntries calls the data handler to process a batch of entries, and calculates & logs the current batch progress.
+func (consumer *StateSyncerConsumer) manageBatchedEntries(batchedEntries []*lib.StateChangeEntry, isBatchMempool bool, entryCount uint64, batchCount uint64) {
+	// Call the data handler to process the batch. We do this with retries, in case the data handler fails.
+	err := consumer.callHandlerWithRetries(batchedEntries, 0)
+	if err != nil {
+		glog.Fatalf("consumer.manageBatchedEntries: %v", err)
```

Comment:
Usually when you find yourself needing to separate the Lock() from the Unlock(), it's best to break the logic in between into a separate function and use defer. You can also use an inner function. Looks like this (and always explain with a comment since it's not a common pattern):

```
	// Prevent multiple threads from accessing the batch index slice at the same time.
	// Use an inner function to unlock the mutex with a defer statement.
	func(){
		consumer.ThreadMutex.Lock()
		defer consumer.ThreadMutex.Unlock()
		// Upon success, add the batch index info to the batch index slice.
		batchInfo := &BatchIndexInfo{
			MinEntryIndex: entryCount + uint64(len(batchedEntries)),
			Index:         batchCount,
		}
		batchIndexes := consumer.BatchIndexes
		consumer.BatchIndexes = insertBatchIndexInOrder(batchIndexes, batchInfo)

		// Log progress.
		fmt.Printf("Handled batch %d\n", batchCount)

		// If the number of batches is greater than the thread limit, remove the first batch index from the slice.
		// We know that anything outside the bounds of the thread limit must have already been processed successfully.
		if len(consumer.BatchIndexes) > consumer.ThreadLimit {
			consumer.BatchIndexes = consumer.BatchIndexes[1:]
		}

		lastConsecutiveBatchEntryIndex := consumer.findLastConsecutiveBatchEntryIndex()

		if !isBatchMempool {
			// Save the last consecutive batch entry index to file. This is used to resume from a failed state.
			if err = consumer.saveConsumerProgressToFile(lastConsecutiveBatchEntryIndex); err != nil {
				glog.Errorf("consumer.manageBatchedEntries: %v", err)
			}
		}
	}()

```